clang/test/C/C2y/n3369.c

```diff
@@ -0,0 +1,117 @@
+  static_assert(sizeof(int[m]));        // expected-error {{static assertion expression is not an integral constant expression}}
+  static_assert(_Countof(int[m]));      // expected-error {{static assertion expression is not an integral constant expression}}
+  
+  // Note that this applies to each array dimension.
+  int another_array[n][7];
+  static_assert(_Countof(another_array)); // expected-error {{static assertion expression is not an integral constant expression}}
+  static_assert(_Countof(*another_array) == 7);
+
+  // Only the first dimension is needed for constant evaluation; other
+  // dimensions can be ignored.
+  int yet_another_array[7][n];
+  static_assert(_Countof(yet_another_array) == 7);
+  static_assert(_Countof(*yet_another_array)); // expected-error {{static assertion expression is not an integral constant expression}}
+  
+  int one_more_time[n][n][7];
+  static_assert(_Countof(one_more_time));  // expected-error {{static assertion expression is not an integral constant expression}}
+  static_assert(_Countof(*one_more_time)); // expected-error {{static assertion expression is not an integral constant expression}}
+  static_assert(_Countof(**one_more_time) == 7);
+}
+
+void test_with_function_param(int array[12], int (*array_ptr)[12], int static_array[static 12]) {
+  (void)_Countof(array); // expected-error {{'_Countof' requires an argument of array type; 'int *' invalid}}
+  static_assert(_Countof(*array_ptr) == 12);
+  (void)_Countof(static_array); // expected-error {{'_Countof' requires an argument of array type; 'int *' invalid}}
+}
+
+void test_multidimensional_arrays() {
+  int array[12][7];
+  static_assert(_Countof(array) == 12);
+  static_assert(_Countof(*array) == 7);
+
+  int mdarray[12][7][100][3];
+  static_assert(_Countof(mdarray) == 12);
+  static_assert(_Countof(*mdarray) == 7);
+  static_assert(_Countof(**mdarray) == 100);
+  static_assert(_Countof(***mdarray) == 3);
+}
+
+void test_unspecified_array_length() {
+  static_assert(_Countof(int[])); // expected-error {{invalid application of '_Countof' to an incomplete type 'int[]'}}
+
+  extern int x[][6][3];
+  static_assert(_Countof(x)); // expected-error {{invalid application of '_Countof' to an incomplete type 'int[][6][3]'}}
+  static_assert(_Countof(*x) == 6);
+  static_assert(_Countof(**x) == 3);
+}
+
+// Test that the return type of _Countof is what you'd expect (size_t).
+void test_return_type() {
+  static_assert(_Generic(typeof(_Countof global_array), typeof(sizeof(0)) : 1, default : 0));
+}
+
+// Test that _Countof is able to look through typedefs.
+void test_typedefs() {
+  typedef int foo[12];
+  foo f;
+  static_assert(_Countof(foo) == 12);
+  static_assert(_Countof(f) == 12);
+
+  // Ensure multidimensional arrays also work.
+  foo x[100];
+  static_assert(_Generic(typeof(x), int[100][12] : 1, default : 0));
+  static_assert(_Countof(x) == 100);
+  static_assert(_Countof(*x) == 12);
+}
```

Comment:
What about arrays that are members of a struct? e.g.:

```c
struct A {
  int arr[5];
};
```

Assuming that this won't work on a FAM?

Zero sized arrays?

Reply:
Good idea, I've added more tests for those.

Reply:
> What about arrays that are members of a struct? e.g.:
> 
> ```c
> struct A {
>   int arr[5];
> };
> ```
> 
> Assuming that this won't work on a FAM?
> 
> Zero sized arrays?

I can confirm that this won't work on a FAM.  My implementation for GCC has tests to make sure that `_Countof(p->fam)` results in a compile-time error.